### PR TITLE
BlendModes: Added `blend*` prefix

### PR DESCRIPTION
**Pull request by sunag**
Related issue: https://github.com/mrdoob/three.js/pull/29879#issuecomment-2476499445

**Description**

- [x] [Rename BlendMode -> BlendModes](https://github.com/mrdoob/three.js/commit/12ead7fc8e1b3a2104c79aa314734531b979257f)
- [x] [Added blend prefix](https://github.com/mrdoob/three.js/commit/9ee033cddf5556a6aab30dc2c8eaed44803c420a)
- [x] [Update examples](https://github.com/mrdoob/three.js/commit/9678baa5e18e6830061de417ab353a46d473aea9)

### CODE REVIEW

File: src/nodes/display/BlendModes.js [hunk 1]
```diff
@@ -65,3 +65,33 @@ export const blendNormal = /*@__PURE__*/ Fn( ( [ base, blend ] ) => {
 		{ name: 'blend', type: 'vec4' }
 	]
 } );
+
+// deprecated
+
+export const burn = ( ...params ) => { // @deprecated, r171
+
+	console.warn( 'THREE.TSL: "burn" has been deprecated. Use "blendBurn" instead.' );
```
Comment (WestLangley) [inline src/nodes/display/BlendModes.js:73]: How about
```
"burn" has been renamed. Use "blendBurn" instead.
```

**Comment by sunag** [comment]
Would it be better to rename `blendNormal` to `blendColor`? 
`blendNormals` should be intended for blend geometry normals.

**Comment by Makio64** [comment]
> Would it be better to rename `blendNormal` to `blendColor`? `blendNormals` should be intended for blend geometry normals.

Sorry for my lack of knowlegdes but you mean there is a way to blend the geometries normals without blending the color (output) ? Do you have a usecase in mind you can share to me to understand why peoples will do it ?

**Comment by sunag** [comment]
> Sorry for my lack of knowlegdes but you mean there is a way to blend the geometries normals without blending the color (output) ? Do you have a usecase in mind you can share to me to understand why peoples will do it ?

A common case is mixing two normal maps.

**Comment by WestLangley** [comment]
>Would it be better to rename blendNormal to blendColor?

Yes, rename it. The question remains as to what the name should be...

The formula looks similar to the Porter Duff _over_ operator, but not exactly.

@Mugen87 Is it an SSR hack of some sort, and therefore specific to that particular algorithm?

//

Be aware that the output of a beauty pass has alpha premultiplied by default. Be careful when using these formulas.

These blend modes only make sense if you know the context. I think the file should have specific references for the formulas.

**Comment by Mugen87** [comment]
> @Mugen87 Is it an SSR hack of some sort, and therefore specific to that particular algorithm?

The previous `SSRPass` used a separate render pass via copy shader to blend the SSR over the beauty via normal blending.  Via `blendColor()` we achieve the same result as before but without a separate pass.

Right now, the blend function is not specific for `SSRNode`. It should work for similar use cases which relied on normal blending as well.

The premultiply alpha topic might be relevant in the future. If a use case comes up, we probably need a third parameter `premultipliedAlpha` with an additional code path.

**Comment by WestLangley** [comment]
@Mugen87

`BlendModes.js` contains some blending functions -- from somewhere -- maybe @sunag knows where they are from. I am curious to know...

>It should work for similar use cases which relied on normal blending as well.

The function you added to the file in #29879 does not match the three.js normal blending formula. Also, the proposed name `blendColor()` does not have a well-defined meaning.

So, currently, I would be inclined to revert #29879, and inline the function in the SSR shader.

**Comment by Mugen87** [comment]
> The function you added to the file in https://github.com/mrdoob/three.js/pull/29879 does not match the three.js normal blending formula.

Do you mind elaborating the differences and why we still get the same visual result?

**Comment by sunag** [comment]
I don't remember exactly all links, because I was looking for blend modes that used `mix` instead of conditionals like this:
https://github.com/Experience-Monks/glsl-blend-overlay/blob/master/index.glsl

then validated them in the image editor and got the expected result.

It looks the "same" comparing to the reference below, except that our version returns a vec4() with the base opacity and has a opacity as parameter, I don't think the order of addition should change the result here.

https://github.com/jamieowen/glsl-blend/blob/master/normal.glsl

**Comment by WestLangley** [comment]
>Do you mind elaborating the differences and why we still get the same visual result?

Please refer to the Porter Duff __over__ blending operator explained in [this Wikipedia article](https://en.wikipedia.org/wiki/Alpha_compositing).

1. The original blending formula assumes the source, destination, and output colors are _non-premultiplied_. Here is how those formulas are written in the case where the source is the shader output and the destination is the drawing buffer -- both non-premultiplied.

```
dst.rgb = src.rgb * src.a + dst.rgb * dst.a * ( 1 - src.a )
dst.a   = src.a   * 1     + dst.a           * ( 1 - src.a )
dst.rgb = dst.rgb / dst.a
```

2. Later in that article, you will find the formula to use when the source, destination, and output colors are premultiplied. This formula is akin to the three.js `NormalBlending` formula when `material.premultipliedAlpha` is `true`.

```
dst.rgb = src.rgb * 1 + dst.rgb * ( 1 - src.a )
dst.a   = src.a   * 1 + dst.a   * ( 1 - src.a )
```

3. Not mentioned in the article is the formula used in three.js for `NormalBlending` when the source (output of the shader) has non-premultiplied alpha and the destination has premultiplied alpha.

```
dst.rgb = src.rgb * src.a + dst.rgb * ( 1 - src.a )
dst.a   = src.a   * 1     + dst.a   * ( 1 - src.a )
```

So, the proper formula to use depends on the context.

You got the "same answer" because `ssr()` ignores the alpha channel of the beauty pass, and assumes alpha is 1. Plus, your test case was likely an opaque one.

//

You have promoted a shader-specific blending formula to a more general status, but the formula is only correct in certain cases. The context is important. Also, the proposed name `blendColor()` does not have a well-defined meaning. This is why I suggested in-lining the formula in the shader, and not promoting it.

**Comment by WestLangley** [comment]
In the renderer's constructor, we set `premultipliedAlpha` to `true`. This is a flag to the compositor that the contents of the drawing buffer have premultiplied alpha.

When working with `WebGPURederer` and TSL, it is important to remember that the input to postprocessing (the output of the beauty pass) will have premultiplied alpha by default. And so should the final output of postprocessing, because that is what the compositor is expecting.

**Comment by sunag** [comment]
I think the problem is in how these formulas are handling opacity, for general use it is better to use the formula that @WestLangley  mentioned (1.), because the SSR blending return a `vec4()` with `base.a` and it would be ignoring the intermediate opacity that `blend.a` could add (It looks like this is a small optimization of instructions for specific SSR usage). All other blending modes are ignoring `opacity` at the moment, but it is very likely that we will need an opacity setting option in the future that works in all blending modes.

**Pull request merged**